pyadjoint/optimization/tao_solver.py

```diff
@@ -131,26 +129,410 @@ def to_petsc(self, x, y):
             x_sub.restoreSubVector(iset, x_sub)
 
 
+def new_control_variable(reduced_functional, *, dual=False):
+    """Return new variables suitable for storing a control value or its dual.
+
+    Args:
+        reduced_functional (ReducedFunctional): The reduced functional whose
+        controls are to be copied.
+        dual (bool): whether to return a dual type. If False then a primal type is returned.
+
+    Returns:
+        tuple[OverloadedType]: New variables suitable for storing a control value.
+    """
+
+    return tuple(control._ad_init_zero(dual=dual)
+                 for control in reduced_functional.controls)
+
+
+def valid_comm(comm):
+    """
+    Return a valid communicator from a user provided Comm or None.
+
+    Args:
+        comm: Optional[Any[petsc4py.PETSc.Comm,mpi4py.MPI.Comm]]
+
+    Returns:
+        mpi4py.MPI.Comm. COMM_WORLD if `comm is None`, otherwise `comm.tompi4py()`.
+    """
+    if comm is None:
+        comm = PETSc.COMM_WORLD
+    if hasattr(comm, "tompi4py"):
+        comm = comm.tompi4py()
+    return comm
+
+
+class RFOperation(Enum):
+    """
+    The type of linear action that a ReducedFunctionalMat should apply.
+    """
+    FORWARD = 'forward'
+    TLM = 'tlm'
+    ADJOINT = 'adjoint'
+    HESSIAN = 'hessian'
+
+
+class ReducedFunctionalMatBase:
+    """
+    PETSc.Mat Python context to apply the action of a pyadjoint.ReducedFunctional.
+
+    If V is the control space and U is the functional space, each action has the following map:
+    Jhat : V -> U
+    TLM : V -> U
+    Adjoint : U* -> V*
+    Hessian : V x U* -> V* | V -> V*
+
+    Args:
+        rf (ReducedFunctional): Defines the forward model, and used to compute operator actions.
+        action (RFOperation): Whether to apply the TLM, adjoint, or Hessian action.
```

Comment:
Wasn't this discussed in the meeting and this was determined a confusing name?

Reply:
Yes but the option is still called `action`. Though see @jrmaddison's comment that this can probably just go away.

Reply:
See response to that comment.